test/jb_splits_store/set.test.js

```diff
@@ -0,0 +1,291 @@
+  const PROJECT_ID = 1;
+  const DOMAIN = 2;
+  const GROUP = 3;
+  let SET_SPLITS_PERMISSION_INDEX;
+
+  before(async function () {
+    let jbOperationsFactory = await ethers.getContractFactory('JBOperations');
+    let jbOperations = await jbOperationsFactory.deploy();
+    SET_SPLITS_PERMISSION_INDEX = await jbOperations.SET_SPLITS();
+  });
+
+  async function setup() {
+    let [deployer, projectOwner, ...addrs] = await ethers.getSigners();
+
+    let mockOperatorStore = await deployMockContract(deployer, jbOperatorStore.abi);
+    let mockProjects = await deployMockContract(deployer, jbProjects.abi);
+    let mockDirectory = await deployMockContract(deployer, jbDirectory.abi);
+
+    await mockOperatorStore.mock.hasPermission
+      .withArgs(projectOwner.address, projectOwner.address, PROJECT_ID, SET_SPLITS_PERMISSION_INDEX)
+      .returns(true);
+
+    await mockProjects.mock.ownerOf
+      .withArgs(PROJECT_ID)
+      .returns(projectOwner.address);
+
+    await mockDirectory.mock.controllerOf
+      .withArgs(PROJECT_ID)
+      .returns(projectOwner.address);
+
+    let jbSplitsStoreFact = await ethers.getContractFactory('JBSplitsStore');
+    let jbSplitsStore = await jbSplitsStoreFact.deploy(
+      mockOperatorStore.address,
+      mockProjects.address,
+      mockDirectory.address
+    );
+
+    let splits = createSplitArray(addrs[0].address, 4);
+
+    return { deployer, projectOwner, addrs, jbSplitsStore, splits, mockOperatorStore, mockProjects, mockDirectory };
+  }
+
+  // Create array of JBSplit struct
+  function createSplitArray(beneficiaryAddress, n) {
+    let splits = []
+    for (let i = 0; i < n; i++) {
+      splits.push({
+        preferClaimed: false,
+        percent: Math.floor(10000000 / n),
+        lockedUntil: 0,
+        beneficiary: beneficiaryAddress,
+        allocator: ZERO_ADDRESS,
+        projectId: 0
+      });
+    }
+    return splits;
+  }
+
+  it('set(...) and corresponding events', async function () {
+    const { projectOwner, addrs, jbSplitsStore, splits, mockOperatorStore, mockDirectory } = await setup();
+
+    await mockOperatorStore.mock.hasPermission.returns(false);
+    await mockDirectory.mock.controllerOf.returns(addrs[0].address);
+
+    const tx = await jbSplitsStore.connect(projectOwner).set(
+      PROJECT_ID,
+      DOMAIN,
+      GROUP,
+      splits
+    );
+
+    // Expect one event per split in splits[]
+    await expect(tx)
+      .to.emit(jbSplitsStore, 'SetSplit')
+      .withArgs(PROJECT_ID, DOMAIN, GROUP, Object.values(splits[0]), projectOwner.address)
+      .and.to.emit(jbSplitsStore, 'SetSplit')
+      .withArgs(PROJECT_ID, DOMAIN, GROUP, Object.values(splits[1]), projectOwner.address)
+      .and.to.emit(jbSplitsStore, 'SetSplit')
+      .withArgs(PROJECT_ID, DOMAIN, GROUP, Object.values(splits[2]), projectOwner.address)
+      .and.to.emit(jbSplitsStore, 'SetSplit')
+      .withArgs(PROJECT_ID, DOMAIN, GROUP, Object.values(splits[3]), projectOwner.address)
```

Comment:
We should probably create an expectation for each event so we don't end up with magic indexes. Consider something like:

    await Promise.all(
      splits.map(async (split, _) => {
        await expect(tx)
          .to.emit(jbSplitsStore, 'SetSplit')
          .withArgs(PROJECT_ID, DOMAIN, GROUP, Object.values(splits[3]), projectOwner.address);
      }),
    );

Reply:
And it's definitely more elegant!